### PR TITLE
Add option to run all tests

### LINKED ISSUE #50363: libtest: provide a way to run *all* tests
I originally [raised this against Cargo](https://github.com/rust-lang/cargo/issues/5419), but I've since learned that this is to do with `libtest` inside Rust itself.

So we have:
```
cargo test
```

To run all but ignored tests. Then:
```
cargo test -- --ignored
```

To run only ignored tests.

But as far as I can see there is no way to run **all** tests in parallel short of:
```
cargo test & cargo test -- --ignored
```

Which isn't ideal for scripting and is not convenient.

How about?
```
cargo test -- --all
```

Cheers

### PR DESCRIPTION
**Pull request by Emerentius**
This adds the "--include-ignored" flag to libtest, which allows running ignored and unignored tests in one go.

Closes #50363

### CODE REVIEW

File: src/libtest/lib.rs [hunk 1]
```diff
@@ -544,7 +552,11 @@ pub fn parse_opts(args: &[String]) -> Option<OptRes> {
         None
     };
 
-    let run_ignored = matches.opt_present("ignored");
+    let run_ignored = match (matches.opt_present("all"), matches.opt_present("ignored")) {
+        (true, _) => RunIgnored::Yes,
```
Comment (Mark-Simulacrum) [inline src/libtest/lib.rs]: I wonder if it makes sense to error here if we've been passed both; what do you think?

**Comment by Emerentius** [comment]
I don't know if "--all" is such a good name when we also have flags for "--tests" and "--benches", but I'm pulling blank on alternatives. I'm thinking it should be ok because the most common usage would be `cargo test -- --all`.

File: src/libtest/lib.rs [hunk 2]
```diff
@@ -1869,7 +1887,21 @@ mod tests {
             Some(Ok(o)) => o,
             _ => panic!("Malformed arg in parse_ignored_flag"),
         };
-        assert!((opts.run_ignored));
+        assert!((opts.run_ignored == RunIgnored::Only));
```
Comment (Mark-Simulacrum) [inline src/libtest/lib.rs]: assert_eq and needless inner parentheses

File: src/libtest/lib.rs [hunk 3]
```diff
@@ -1869,7 +1887,21 @@ mod tests {
...
+    #[test]
+    fn parse_all_flag() {
+        let args = vec![
+            "progname".to_string(),
+            "filter".to_string(),
+            "--all".to_string(),
+        ];
+        let opts = match parse_opts(&args) {
+            Some(Ok(o)) => o,
+            _ => panic!("Malformed arg in parse_ignored_flag"),
```
Comment (Mark-Simulacrum) [inline src/libtest/lib.rs]: No need for this panic, we can probably just unwrap?

**Review by Mark-Simulacrum — DISMISSED** [review]
A few minor nits, but otherwise this looks good.

**Comment by Mark-Simulacrum** [comment]
@rust-lang/libs - as an insta-stable change asking for FCP (I don't think I can start it).

Cc @rust-lang/dev-tools, not sure if this is under your purview...

**Comment by SimonSapin** [comment]
libtest is semi-private / permanently unstable (except possibly for `Bencher`), so I don’t think there is a library discussion to be had here.

File: src/libtest/lib.rs (hunk 3 above)
Comment (Emerentius) [reply src/libtest/lib.rs]: Yeah. Also, I think this is an error message for a buggy test, not a buggy parse_opts.

File: src/libtest/lib.rs (hunk 1 above)
Comment (Emerentius) [reply src/libtest/lib.rs]: Hmm, maybe. It's often better for humans but worse for automated tools.

**Comment by Manishearth** [comment]
@djrenren on devtools is working on custom test frameworks and probably should have a say on this

**Comment by djrenren** [comment]
This looks really helpful. Two nits:

 - I'd recommend `--force` that way this could be used with test, bench, or filtered subset. It also conveys that we'd be overriding something.

 - Since this is a change to the public-facing interface (which I believe we consider stable?) shouldn't we run this through nightly?

**Comment by Emerentius** [comment]
I associate `--force` quite strongly, though not exclusively, with operations that are dangerous, inadvisable or rarely needed like `rm -f` or `git {push | pull | clean | checkout } --force`


> shouldn't we run this through nightly?

What does that mean in this context?

**Comment by djrenren** [comment]
Presumably ignored tests are ignored for a reason. You're explicitly going against the intentions of the author so to me it seems worth of a `--force`, but maybe I don't have the correct mental model for ignored tests.

> What does that mean in this context?

Ah @Mark-Simulacrum mentioned making this an insta-stable change rather than starting as nightly-only then stabilizing later.

**Comment by Emerentius** [comment]
If the author didn't want it to be run ever it would be deleted, commented out or not marked `#[test]`. There is of course no prescribed reason for which `#[ignore]` may be used, but realistic cases would be for example very expensive computations or things that access external resources (say, access certain files that may not be setup). In fact, expensive computations is what I've used them for and it's also is what the book uses as example. A test in cargo's test suite is ignored because it does a web request.

**Comment by SimonSapin** [comment]
It isn’t great that the subtly different commands `cargo test --all` and `cargo test -- --all` would do very different things. I’d prefer another name, but I don’t really have a suggestion.

**Comment by nrc** [comment]
Seems like a nice thing to have from my perspective. I agree `--all` is not great. Maybe `--test-ignored` or something similarly explicit?

**Comment by Emerentius** [comment]
--include-ignored ?

**Comment by Mark-Simulacrum** [comment]
I'm going to reassign to r? @nrc since I think  the dev tools team is where the decision needs to come from at this point

**Comment by nrc** [comment]
We'll discuss at this weeks meeting.

**Comment by nrc** [comment]
We discussed this in the dev-tools meeting today. We mostly felt positive about the change. Although we were kind of unsatisfied, we all think `--include-ignored` is the best option for the argument.

As a kind of 'poor man's feature gate', could we have this feature only turned on for nightly/dev Rust and not for beta and stable? (Insta-stable always feels bad).

**Comment by Emerentius** [comment]
> As a kind of 'poor man's feature gate', could we have this feature only turned on for nightly/dev Rust and not for beta and stable? 

libtest should check nightly status, but it's apparently broken. I can use `-Zunstable-options` on stable although it says, that it's nightly only.  
The `is_nightly` function is small and seems reasonable, but I don't know at what point libtest is compiled and with what settings.

```rust
fn is_nightly() -> bool {
    // Whether this is a feature-staged build, i.e. on the beta or stable channel
    let disable_unstable_features = option_env!("CFG_DISABLE_UNSTABLE_FEATURES").is_some();
    // Whether we should enable unstable features for bootstrapping
    let bootstrap = env::var("RUSTC_BOOTSTRAP").is_ok();

    bootstrap || !disable_unstable_features
}
```

**Comment by Emerentius** [comment]
@nrc I have renamed the option and gated it on the presence of `-Zunstable-options`.

Like I said above, `-Zunstable-option`'s limitation to nightly is broken, but it still clearly communicates that `include-ignored` isn't a stable option.

**Comment by nrc** [comment]
@bors: r+

Thanks for the changes!

**Comment by kennytm** [comment]
This has broken a dependency `compiletest_rs` which clippy's test relies on

```
[00:55:31] error[E0308]: mismatched types
[00:55:31]    --> C:\Users\appveyor\.cargo\registry\src\github.com-1ecc6299db9ec823\compiletest_rs-0.3.13\src\lib.rs:100:22
[00:55:31]     |
[00:55:31] 100 |         run_ignored: config.run_ignored,
[00:55:31]     |                      ^^^^^^^^^^^^^^^^^^ expected enum `test::RunIgnored`, found bool
[00:55:31]     |
[00:55:31]     = note: expected type `test::RunIgnored`
[00:55:31]                found type `bool`
[00:55:31] 
[00:55:32] error: aborting due to previous error
```

**Comment by Emerentius** [comment]
@kennytm So that's an external crate that calls into rustc internals? I'm a bit confused on what I should do now.
Get a patch ready for `compiletest_rs` so this can be changed in sync? Some intermediate breakage seems necessary.

**Comment by Mark-Simulacrum** [comment]
You'll probably want to just retry the build in a few days at which point this should land since we don't require clippy to pass normally, just before and during release week. Once this lands compiletest on crates.io can be updated and finally clippy update can land (or just Cargo.lock, perhaps).

**Comment by eddyb** [comment]
In the interest of saving time, and breaking as little as possible, I'll add a `From<bool>` impl for this new `enum` in #54116, and update `compiletest-rs` to use `From::from`.

**Comment by eddyb** [comment]
So I think we should try to do this *without* breaking `compiletest_rs`. I'm looking into it right now.

**Comment by Emerentius** [comment]
@eddyb Have the waters calmed enough that this can be merged again? Should I open a new PR?

**Comment by eddyb** [comment]
@Emerentius Yes, sorry to delay this and keep you waiting! RC1 went out about a week ago.

**Pull request merged**